compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Parsing.hs

```diff
@@ -0,0 +1,188 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE ApplicativeDo       #-}
+{-# LANGUAGE RankNTypes       #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+
+module DA.Cli.Damlc.Command.MultiIde.Parsing (
+  getChunks,
+  parseClientMessageWithTracker,
+  parseServerMessageWithTracker,
+  putChunk,
+  putReqMethodAll,
+  putReqMethodSingleFromClient,
+  putReqMethodSingleFromServer,
+  putServerReq,
+) where
+
+import Control.Concurrent.STM.TVar
+import Control.Lens
+import Control.Monad.STM
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.Attoparsec.ByteString.Lazy as Attoparsec
+import qualified Data.ByteString as B
+import Data.ByteString.Builder.Extra (defaultChunkSize)
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.ByteString.Lazy.Char8 as BSLC
+import Data.Foldable (forM_)
+import DA.Cli.Damlc.Command.MultiIde.Types
+import Data.Functor.Product
+import qualified Data.IxMap as IM
+import Data.List (delete)
+import Data.Maybe (fromMaybe)
+import qualified Language.LSP.Types as LSP
+import System.IO.Extra
+import System.IO.Unsafe (unsafeInterleaveIO)
+
+
+{-# ANN allBytes ("HLint: ignore Avoid restricted function" :: String) #-}
+-- unsafeInterleaveIO used to create a chunked lazy bytestring over IO for a given Handle
+allBytes :: Handle -> IO BSL.ByteString
+allBytes hin = fmap BSL.fromChunks go
+  where
+    go :: IO [B.ByteString]
+    go = do
+      first <- unsafeInterleaveIO $ B.hGetSome hin defaultChunkSize
+      rest <- unsafeInterleaveIO go
```

Comment:
Possibly we should take this out, as clever as it is :angel: :innocent: , lsp already has a function to make this look ugly but safer
Sam: considering you wrote this code, this monster is for you

Reply:
Adding a TODO for ya :)